### PR TITLE
Implement nested folder support and improve file operations on pCloud Backend

### PR DESCRIPTION
**Pull request by marceloduplicati**
- Add support for subfolders
- Implement robust file selection using fileId for Download/Delete operations instead of concatenation of path + filename and folder selection using folderId rather than pathnames.

This is a fix for the issue raised on the forum: https://forum.duplicati.com/t/pcloud-and-duplicati/8546/16

### CODE REVIEW

File: Duplicati/Library/Backend/pCloud/pCloudBackend.cs [hunk 1]
```diff
@@ -236,10 +250,8 @@ public async Task PutAsync(string remotename, string localname, CancellationToke
     /// <exception cref="Exception">Exceptions arising from either code execution</exception>
     public async Task PutAsync(string remotename, Stream input, CancellationToken cancellationToken)
     {
-        // If we don't have the folder ID cached, lets fetch it with the create folder api
-        if (_CachedFolderID == null)
-            await CreateFolderAsync(cancellationToken).ConfigureAwait(false);
-
+        _CachedFolderID ??= GetFolderId().Await();
```
Comment (kenkendk) [inline Duplicati/Library/Backend/pCloud/pCloudBackend.cs:253]: This should be `_CachedFolderID ??= await GetFolderId(cancellationToken).ConfigureAwait(false)`

File: Duplicati/Library/Backend/pCloud/pCloudBackend.cs [hunk 2]
```diff
@@ -341,7 +351,7 @@ public async Task GetAsync(string remotename, Stream output, CancellationToken c
 
             using var requestResources = CreateRequest(string.Empty, HttpMethod.Get);
 
-            requestResources.RequestMessage.RequestUri = new Uri(await GetFileId(remotename));
+            requestResources.RequestMessage.RequestUri = new Uri(GetFileLink(remotename).Await());
```
Comment (kenkendk) [inline Duplicati/Library/Backend/pCloud/pCloudBackend.cs:354]: This should be:
```
 requestResources.RequestMessage.RequestUri = new Uri(await GetFileLink(remotename, cancellationToken).ConfigureAwait(false));
```

File: Duplicati/Library/Backend/pCloud/pCloudBackend.cs [hunk 3]
```diff
@@ -377,15 +387,12 @@ await requestResources.HttpClient.DownloadFile(requestResources.RequestMessage,
...
-        if (!String.IsNullOrWhiteSpace(_Path))
-            remotename = $"/{_Path}/{remotename}";
-
         using var timeoutToken = new CancellationTokenSource();
         timeoutToken.CancelAfter(TimeSpan.FromSeconds(SHORT_OPERATION_TIMEOUT_SECONDS));
         using var combinedTokens =
             CancellationTokenSource.CreateLinkedTokenSource(timeoutToken.Token, cancellationToken);
 
-        using var requestResources = CreateRequest($"/deletefile?path={remotename}", HttpMethod.Get);
+        using var requestResources = CreateRequest($"/deletefile?fileid={GetFileId(remotename).Await()}", HttpMethod.Get);
```
Comment (kenkendk) [inline Duplicati/Library/Backend/pCloud/pCloudBackend.cs:395]: This should be:
```
using var requestResources = CreateRequest($"/deletefile?fileid={await GetFileId(remotename, cancellationToken).ConfigureAwait(false)}", HttpMethod.Get);
```

File: Duplicati/Library/Backend/pCloud/pCloudBackend.cs [hunk 4]
```diff
@@ -459,21 +493,48 @@ public Task CreateFolderAsync(CancellationToken cancellationToken)
...
-        return Task.CompletedTask;
+    /// <summary>
+    /// Returns the fileID by listing the folder and searching for the filename & metadata.
+    ///
+    /// For operations such as delete/getfilelink using the fileID is more reliable than using direct path/filename concatenation.
+    /// </summary>
+    /// <param name="name">The filename</param>
+    /// <returns></returns>
+    /// <exception cref="FileNotFoundException"></exception>
+    private Task<ulong> GetFileId(string name)
```
Comment (kenkendk) [inline Duplicati/Library/Backend/pCloud/pCloudBackend.cs:510]: This should take a cancellationToken

File: Duplicati/Library/Backend/pCloud/pCloudBackend.cs [hunk 5]
```diff
@@ -459,21 +493,48 @@ public Task CreateFolderAsync(CancellationToken cancellationToken)
...
+    {
+        _CachedFolderID ??= GetFolderId().Await();
```
Comment (kenkendk) [inline Duplicati/Library/Backend/pCloud/pCloudBackend.cs:512]: This should be:
```
_CachedFolderID ??= await GetFolderId(cancellationToken).ConfigureAwait(false);
```

File: Duplicati/Library/Backend/pCloud/pCloudBackend.cs [hunk 6]
```diff
@@ -459,21 +493,48 @@ public Task CreateFolderAsync(CancellationToken cancellationToken)
...
+        return Task.FromResult(ListWithMetadata((ulong)_CachedFolderID)
+            .FirstOrDefault(x => !x.isfolder && x.name == name)
+            ?.fileid ?? throw new FileNotFoundException(name));
     }
+    
+    /// <summary>
+    /// Returns the folder ID for the configured path, regardless of the depth
+    /// </summary>
+    /// <exception cref="FolderMissingException"></exception>
+    private Task<ulong> GetFolderId()
```
Comment (kenkendk) [inline Duplicati/Library/Backend/pCloud/pCloudBackend.cs:523]: This should take a cancellation token.

File: Duplicati/Library/Backend/pCloud/pCloudBackend.cs [hunk 7]
```diff
@@ -155,19 +167,31 @@ public IList<ICommandLineArgument> SupportedCommands
...
+            .ToList();
+    }
+    
+    /// <summary>
+    /// Lists folders with pCloud metadata, necessary to obtain the folder IDs
+    /// </summary>
+    /// <param name="folderId">FolderId to be used as root.</param>
+    /// <returns></returns>
+    /// <exception cref="Exception"></exception>
+    private IEnumerable<pCloudFolderContent> ListWithMetadata(ulong folderId)
```
Comment (kenkendk) [inline Duplicati/Library/Backend/pCloud/pCloudBackend.cs:193]: Consider if this should be `async` and return `Task<List<pCloudFolderContent>>` so we can pass in a cancellation token.

File: Duplicati/Library/Backend/pCloud/pCloudBackend.cs [hunk 8]
```diff
@@ -418,17 +425,23 @@ public Task DeleteAsync(string remotename, CancellationToken cancellationToken)
...
-        if (!List(true).Any(x => x.IsFolder && x.Name == _Path) && !string.IsNullOrWhiteSpace(_Path))
-            throw new FolderMissingException();
+        if (string.IsNullOrWhiteSpace(_Path))
+            return Task.CompletedTask;
+
+        var segments = _Path.Split(PATH_SEPARATORS, StringSplitOptions.RemoveEmptyEntries);
+        if (segments.Length == 0)
+            return Task.CompletedTask;
+
+        _ = segments.Aggregate(0UL, (parentId, folder) =>
```
Comment (kenkendk) [inline Duplicati/Library/Backend/pCloud/pCloudBackend.cs:441]: Not convinced this is not skipped by the runtime. The result is never enumerated, sot it likely never calls the method.

**Review by kenkendk — APPROVED** [review]

File: Duplicati/Library/Backend/pCloud/pCloudBackend.cs (hunk 8 above)
Comment (marceloduplicati) [reply Duplicati/Library/Backend/pCloud/pCloudBackend.cs]: This was substantially changed on the new commit when CancellationTokens were propagated, its not using Aggregate anymore.

File: Duplicati/Library/Backend/pCloud/pCloudBackend.cs (hunk 3 above)
Comment (marceloduplicati) [reply Duplicati/Library/Backend/pCloud/pCloudBackend.cs]: Latest commit has implemented this alongside with all cancellation token propagation.

File: Duplicati/Library/Backend/pCloud/pCloudBackend.cs (hunk 6 above)
Comment (marceloduplicati) [reply Duplicati/Library/Backend/pCloud/pCloudBackend.cs]: Latest commit has implemented this alongside with all cancellation token propagation.

File: Duplicati/Library/Backend/pCloud/pCloudBackend.cs (hunk 4 above)
Comment (marceloduplicati) [reply Duplicati/Library/Backend/pCloud/pCloudBackend.cs]: Latest commit has implemented this alongside with all cancellation token propagation.

File: Duplicati/Library/Backend/pCloud/pCloudBackend.cs (hunk 7 above)
Comment (marceloduplicati) [reply Duplicati/Library/Backend/pCloud/pCloudBackend.cs]: Latest commit has implemented this alongside with all cancellation token propagation.

**Review by kenkendk — APPROVED** [review]

**Pull request merged**